### PR TITLE
Add VICReg loss

### PR DESCRIPTION
**Pull request by b-albar**
It adds a VICReg loss that is described in this [paper](https://arxiv.org/abs/2105.04906) also as proposed in #779 
The implementation follows the official code, [here](https://github.com/facebookresearch/vicreg). In term of architecture, it's similar to BarlowTwins, so there is no need for a new projection head.

I would appreciate any feedback on this.

### CODE REVIEW

**Comment by philippmwirth** [comment]
Thanks for the contribution @b-albar! We'll take a look shortly 🙂

File: lightly/loss/vicreg_loss.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    def forward(self, z_a: torch.Tensor, z_b: torch.Tensor) -> torch.Tensor:
+
+        # invariance term of the loss
+        repr_loss = F.mse_loss(z_a, z_b)
+
+        # gather all batches
+        if self.gather_distributed and dist.is_initialized():
+            world_size = dist.get_world_size()
+            if world_size > 1:
+                dist.all_reduce(z_a)
```
Comment (philippmwirth) [inline lightly/loss/vicreg_loss.py:69]: I believe this should be `all_gather` rather than `all_reduce`. What do you think @b-albar  and @guarin?

**Review by philippmwirth — COMMENTED** [review]
Thanks a lot for the contribution! I think once we fix the distributed gathering we can merge it 🙂 

@b-albar please let us know if you want to do it yourself or if you want someone to take over the PR.

File: lightly/loss/vicreg_loss.py (hunk 1 above)
Comment (b-albar) [reply lightly/loss/vicreg_loss.py:69]: Yes I agree it should be all_gather. But in this case, I think it should use the GatherLayer in lightly.utils.dist to avoid breaking the gradient flow during backprop. I'll fix the PR.

**Review by guarin — APPROVED** [review]

**Review by guarin — CHANGES_REQUESTED** [review]
Hi, thanks a lot for the PR and for fixing the gathering operation!

We had some issues with the CI tests not running through because setuptools was missing. This has now been fixed on the master branch. I just tried your code with the updates from the master branch but now the vicreg tests fail because `nan` is returned.

New branch I made: https://github.com/lightly-ai/lightly/pulls
Failing tests: https://github.com/lightly-ai/lightly/actions/runs/3530335397/jobs/5922217442

Do you know what the issue for the nan could be? And could you also merge our master branch back into your branch to apply the latest changes?

**Comment by guarin** [comment]
Hi @b-albar I managed to fix the bug in the tests (batch size must be at least 2) and merged the changes in #989. Thanks a ton for your contribution!
